### PR TITLE
support bloom filter

### PR DESCRIPTION
**Pull request by hero5512**

### CODE REVIEW

File: common/config/config.go [hunk 1]
```diff
@@ -261,6 +261,18 @@ func GetAddDecimalsHeight() uint32 {
 	}
 }
 
+// TODO mainnet forking blockNumber
+func GetAddFilterHeight() uint32 {
```
Comment (laizy) [inline common/config/config.go]: why need hardfork? this feature do not have any consensus changes.

File: core/ledger/ledger.go [hunk 2]
```diff
@@ -39,6 +40,10 @@ func (self *Ledger) GetStore() store.LedgerStore {
 
 func InitLedger(dataDir string, stateHashHeight uint32, defaultBookkeeper []keypair.PublicKey,
 	genesisBlock *types.Block) (*Ledger, error) {
+	// must init indexer before ledger store
```
Comment (laizy) [inline core/ledger/ledger.go]: more detail why need initialize first.

File: core/store/indexstore/bloombits.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	b.NewBatch()
+	bloomData := make([]*KV, types.BloomBitLength)
+	for i := 0; i < types.BloomBitLength; i++ {
+		bits, err := b.gen.Bitset(uint(i))
+		if err != nil {
+			return nil, nil
+		}
+		value := bitutil.CompressBytes(bits)
+		b.store.BatchPut(bloomBitsKey(uint(i), b.section, b.head), value)
+		bloomData[i] = &KV{Key: bloomBitsKey(uint(i), b.section, b.head), Value: value}
```
Comment (laizy) [inline core/store/indexstore/bloombits.go]: reuse bloomBitsKey generated above?

File: core/store/indexstore/bloombits.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+//CommitTo commit the batch to store
+func (this *bloomIndexer) CommitTo() error {
+	return this.store.BatchCommit()
+}
+
+// bloomBitsKey = bloomBitsPrefix + bit (uint16 big endian) + section (uint32 big endian) + hash
+func bloomBitsKey(bit uint, section uint32, hash common.Hash) []byte {
+	key := append(append(bloomBitsPrefix, make([]byte, 10)...), hash.Bytes()...)
```
Comment (laizy) [inline core/store/indexstore/bloombits.go]: why 10?

File: core/store/indexstore/bloombits.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+func (b *bloomIndexer) Reset(section uint32) error {
+	gen, err := bloombits.NewGenerator(uint(b.size))
+	b.gen, b.section, b.head = gen, section, common.Hash{}
+	return err
+}
+
+// Process implements core.ChainIndexerBackend, adding a new header's bloom into
+// the index.
+func (b *bloomIndexer) Process(hash common.Hash, height uint32, bloom types.Bloom) {
+	// the initial height is 1 but it on ethereum is 0. so subtract 1
```
Comment (laizy) [inline core/store/indexstore/bloombits.go]: where substract 1?

File: core/store/indexstore/indexer.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	}
+	i.storedSections = sections // needed if new > old
+}
+
+// setBloomData put SectionHead and ValidSections into watcher.bloomData
+func (i *Indexer) setBloomData(bloomData *[]*KV, section uint32, hash common.Hash) {
+	var data [8]byte
+	binary.BigEndian.PutUint32(data[:], section)
+	*bloomData = append(*bloomData, &KV{Key: append([]byte("shead"), data[:]...), Value: hash.Bytes()})
+	*bloomData = append(*bloomData, &KV{Key: []byte("count"), Value: data[:]})
```
Comment (laizy) [inline core/store/indexstore/indexer.go]: extract out this keys as constants

File: core/store/indexstore/indexer.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+// GetValidSections reads the number of valid sections from the index database
+// and caches is into the local state.
+func (i *Indexer) GetValidSections() uint32 {
+	data, _ := i.backend.store.Get([]byte("count"))
```
Comment (laizy) [inline core/store/indexstore/indexer.go]: 所有对db的读写都需要包装在backend的对外函数里，类似ReadBloomBits。不要绕过直接读写背后的store。

File: core/store/indexstore/indexer.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+		i.removeSectionHead(i.storedSections)
+	}
+	i.storedSections = sections // needed if new > old
+}
+
+// setBloomData put SectionHead and ValidSections into watcher.bloomData
+func (i *Indexer) setBloomData(bloomData *[]*KV, section uint32, hash common.Hash) {
+	var data [8]byte
+	binary.BigEndian.PutUint32(data[:], section)
+	*bloomData = append(*bloomData, &KV{Key: append([]byte("shead"), data[:]...), Value: hash.Bytes()})
```
Comment (laizy) [inline core/store/indexstore/indexer.go]: where bloomData used?

File: core/store/indexstore/indexer.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+	if len(hash) == len(common.Hash{}) {
+		return common.BytesToHash(hash)
+	}
+	return common.Hash{}
+}
+
+// setSectionHead writes the last block hash of a processed section to the index
+// database.
+func (i *Indexer) setSectionHead(section uint32, hash common.Hash) {
+	var data [8]byte
```
Comment (laizy) [inline core/store/indexstore/indexer.go]: why 8?

File: core/store/indexstore/indexer.go [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+	data, _ := i.backend.store.Get([]byte("count"))
+	if len(data) == 8 {
+		return binary.BigEndian.Uint32(data)
+	}
+	return 0
+}
+
+// sectionHead retrieves the last block hash of a processed section from the
+// index database.
+func (i *Indexer) sectionHead(section uint32) common.Hash {
```
Comment (laizy) [inline core/store/indexstore/indexer.go]: what is section head used for?

File: core/store/indexstore/indexer.go [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+
+		begin := section*BloomBitsBlocks + config.GetAddFilterHeight()
+		end := (section+1)*BloomBitsBlocks + config.GetAddFilterHeight()
+
+		for number := begin; number < end; number++ {
+			var (
+				bloom ethtypes.Bloom
+				hash  common.Hash
+			)
+			i.updateBlock()
```
Comment (laizy) [inline core/store/indexstore/indexer.go]: why need this call?

File: core/store/indexstore/bloombits.go [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+)
+
+// bloomIndexer implements a core.ChainIndexer, building up a rotated bloom bits index
+// for the Ethereum header bloom filters, permitting blazing fast filtering.
+type bloomIndexer struct {
+	size    uint32                     // section size to generate bloombits for
+	store   *leveldbstore.LevelDBStore // database instance to write index data and metadata into
+	gen     *bloombits.Generator       // generator to rotate the bloom bits crating the bloom index
+	section uint32                     // Section is the section number being processed currently
+	head    common.Hash                // Head is the hash of the last header processed
```
Comment (laizy) [inline core/store/indexstore/bloombits.go]: why need this head field?

File: core/store/ledgerstore/ledger_store.go [hunk 13]
```diff
@@ -985,12 +1021,26 @@ func (this *LedgerStoreImp) submitBlock(block *types.Block, crossChainMsg *types
...
+			go func() {
+				indexer.NotifyNewHeight()
+			}()
+		} else {
+			interval := block.Header.Height - config.GetAddFilterHeight()
+			if interval >= (indexer.GetValidSections()+1)*indexstore.BloomBitsBlocks {
+				go indexer.ProcessSection(this, interval)
+			}
+		}
+	}
```
Comment (laizy) [inline core/store/ledgerstore/ledger_store.go]: 这里搞复杂了呀，内存里保存一份bloombits，每生成一个区块就把bloom添加进去，到了4096个就落账。不需要做成异步的，开一堆go routine，还没有任何错误处理。

File: http/ethrpc/filters/filter_system.go [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+	// LastSubscription keeps track of the last index
+	LastIndexSubscription
+)
+
+const (
+	// txChanSize is the size of channel listening to NewTxsEvent.
+	// The number is referenced from the size of tx pool.
+	txChanSize = 4096
+	// rmLogsChanSize is the size of channel listening to RemovedLogsEvent.
+	rmLogsChanSize = 10
```
Comment (laizy) [inline http/ethrpc/filters/filter_system.go]:
```suggestion
```

File: http/ethrpc/filters/filter_system.go [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+	bactor.SubscribeEvent(message.TOPIC_PENDING_TX_EVENT, m.pushPendingTxEvent)
+	bactor.SubscribeEvent(message.TOPIC_ETH_SC_EVENT, m.pushSCEvent)
+	bactor.SubscribeEvent(message.TOPIC_CHAIN_EVENT, m.pushChainEvent)
+
+	//m.txsSub = m.backend.SubscribeNewTxsEvent(m.txsCh)
+	//m.logsSub = m.backend.SubscribeLogsEvent(m.logsCh)
+	//m.rmLogsSub = m.backend.SubscribeRemovedLogsEvent(m.rmLogsCh)
+	//m.chainSub = m.backend.SubscribeChainEvent(m.chainCh)
+	//m.pendingLogsSub = m.backend.SubscribePendingLogsEvent(m.pendingLogsCh)
+
```
Comment (laizy) [inline http/ethrpc/filters/filter_system.go]:
```suggestion
```

File: http/ethrpc/filters/filter_system.go [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+// SubscribeNewHeads creates a subscription that writes the header of a block that is
+// imported in the chain.
+func (es *EventSystem) SubscribeNewHeads(headers chan *types.Header) *Subscription {
+	sub := &subscription{
+		id:        rpc.NewID(),
+		typ:       BlocksSubscription,
+		created:   time.Now(),
+		logs:      make(chan []*types.Log),
+		hashes:    make(chan []common.Hash),
```
Comment (laizy) [inline http/ethrpc/filters/filter_system.go]: those fields should never be used, remove them to make it more clear?

File: http/ethrpc/filters/filter_system.go [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+// SubscribePendingTxs creates a subscription that writes transaction hashes for
+// transactions that enter the transaction pool.
+func (es *EventSystem) SubscribePendingTxs(hashes chan []common.Hash) *Subscription {
+	sub := &subscription{
+		id:        rpc.NewID(),
+		typ:       PendingTransactionsSubscription,
+		created:   time.Now(),
+		logs:      make(chan []*types.Log),
+		hashes:    hashes,
+		headers:   make(chan *types.Header),
```
Comment (laizy) [inline http/ethrpc/filters/filter_system.go]: ditto

File: http/ethrpc/filters/filter_system.go [hunk 18]
```diff
@@ -0,0 +1,330 @@
...
+// given criteria to the given logs channel.
+func (es *EventSystem) subscribeLogs(crit ethereum.FilterQuery, logs chan []*types.Log) *Subscription {
+	sub := &subscription{
+		id:        rpc.NewID(),
+		typ:       LogsSubscription,
+		logsCrit:  crit,
+		created:   time.Now(),
+		logs:      logs,
+		hashes:    make(chan []common.Hash),
+		headers:   make(chan *types.Header),
```
Comment (laizy) [inline http/ethrpc/filters/filter_system.go]: those fields should never be used, remove them to make it more clear?

File: http/ethrpc/filters/filter.go [hunk 19]
```diff
@@ -0,0 +1,482 @@
...
+	if f.criteria.ToBlock.Int64() == -1 {
+		f.criteria.ToBlock = big.NewInt(int64(curHeight))
+	}
+
+	if f.criteria.FromBlock.Int64() <= int64(config.GetAddFilterHeight()) ||
+		f.criteria.ToBlock.Int64() <= int64(config.GetAddFilterHeight()) {
+		return nil, fmt.Errorf("from and to block height must greater than %d", int64(config.GetAddFilterHeight()))
+	}
+
+	// TODO limit search span
```
Comment (laizy) [inline http/ethrpc/filters/filter.go]: do it.

File: core/store/indexstore/bloombits.go [hunk 20]
```diff
@@ -129,9 +123,49 @@ func (this *bloomIndexer) CommitTo() error {
...
+	}
+	return height, nil
+}
+
+func genFilterStartKey() []byte {
+	return []byte{byte(scom.ST_ETH_FILTER_START)}
+}
+
+// setValidSections writes the number of valid sections to the index database
+func (this *bloomIndexer) SetValidSections(sections uint32) error {
```
Comment (laizy) [inline core/store/indexstore/bloombits.go]: why need save this? the section after the start index height will always be available.

File: core/store/indexstore/indexer.go [hunk 21]
```diff
@@ -52,31 +38,40 @@ func CloseIndexer() {
...
-	backend bloomIndexer // Background processor generating the index data content
+	backend        bloomIndexer // Background processor generating the index data content
+	start          uint32       // Start is the start height of bloom filter supported
+	storedSections uint32       // Number of sections successfully indexed into the database
+}
 
-	update chan struct{} // Notification channel that headers should be processed
-	quit   chan struct{} // Quit channel to tear down running goroutines
+func New(dataDir string, curBlockHeight uint32) (*Indexer, error) {
+	db, err := leveldbstore.NewLevelDBStore(filepath.Join(dataDir, bloomIdxDir))
```
Comment (laizy) [inline core/store/indexstore/indexer.go]: no need new db.

File: core/store/indexstore/indexer.go [hunk 22]
```diff
@@ -52,31 +38,40 @@ func CloseIndexer() {
...
 
-func InitIndexer(dataDir string) error {
-	store, err := leveldbstore.NewLevelDBStore(filepath.Join(dataDir, bloomIdxDir))
+	start, err := backend.GetFilterStart()
 	if err != nil {
-		return err
+		if err != common3.ErrNotFound {
+			return nil, fmt.Errorf("new indexer: %s", err.Error())
+		}
+		err = backend.PutFilterStart(curBlockHeight)
```
Comment (laizy) [inline core/store/indexstore/indexer.go]: I do not think we need save all the zero bytes in db before the height evm feature enabled.

File: core/store/indexstore/indexer.go [hunk 23]
```diff
@@ -86,153 +81,86 @@ func (i *Indexer) StoredSection() uint32 {
...
+				return fmt.Errorf("canonical block %d unknown", number)
+			}
+
+			bloom, err := k.GetBloomData(number)
+			if err != nil {
+				return fmt.Errorf("get bloom data height: %d %s", number, err.Error())
 			}
-			i.backend.Process(hash, number, bloom)
-			lastHead = hash
+			i.backend.Process(hash, number, bloom, i.start)
```
Comment (laizy) [inline core/store/indexstore/indexer.go]: why not call this when persist each block so the bloom cache code can be cleaned.

File: core/store/ledgerstore/ledger_store.go [hunk 24]
```diff
@@ -747,6 +779,8 @@ func (this *LedgerStoreImp) saveBlockToBlockStore(block *types.Block) error {
 	if err != nil {
 		return fmt.Errorf("SaveBlock height %d hash %s error %s", blockHeight, blockHash.ToHexString(), err)
 	}
+	this.blockStore.SaveBloomData(blockHeight, bloom)
+	this.setBloomCache(blockHeight, bloom)
```
Comment (laizy) [inline core/store/ledgerstore/ledger_store.go]: update the bloom bits directly, and persist to blockStore if reached 4096.

**Review by laizy — CHANGES_REQUESTED** [review]

File: core/store/indexstore/bloombits.go (hunk 20 above)
Comment (hero5512) [reply core/store/indexstore/bloombits.go]: done

**Pull request merged**